src/Platforms/OMPTarget/OMPallocator.hpp

```diff
@@ -166,5 +168,120 @@ struct qmc_allocator_traits<OMPallocator<T, HostAllocator>>
   }
 };
 
+#if defined(ENABLE_OFFLOAD)
+/** allocator for OMPTarget device memory
+ * @tparam T data type
+ *
+ * using this with something other than Ohmms containers?
+ *  -- use caution, write unit tests! --
+ * It's not tested beyond use in some unit tests using std::vector with constant size.
+ * OMPTargetAllocator appears to meet all the nonoptional requirements of a c++ Allocator.
+ *
+ * Some of the default implementations in std::allocator_traits
+ * of optional Allocator requirements may cause runtime or compilation failures.
+ * They assume there is only one memory space and that the host has access to it.
+ */
+template<typename T>
+class OMPTargetAllocator
+{
+public:
+  using value_type    = T;
+  using size_type     = size_t;
+  using pointer       = T*;
+  using const_pointer = const T*;
+
+  OMPTargetAllocator() = default;
+  template<class U>
+  OMPTargetAllocator(const OMPTargetAllocator<U>&)
+  {}
+
+  template<class U>
+  struct rebind
+  {
+    using other = OMPTargetAllocator<U>;
+  };
+
+  T* allocate(std::size_t n)
+  {
+    void* pt = omp_target_alloc(n * sizeof(T), omp_get_default_device());
+    if(!pt)
+      throw std::runtime_error("Allocation failed in OMPTargetAllocator!");
+    OMPallocator_device_mem_allocated += n * sizeof(T);
+    return static_cast<T*>(pt);
+  }
+
+  void deallocate(T* p, std::size_t n)
+  {
+    omp_target_free(p, omp_get_default_device());
+    OMPallocator_device_mem_allocated -= n * sizeof(T);
+  }
+
+  /** Provide a construct for std::allocator_traits::contruct to call.
+   *  Don't do anything on construct, pointer p is on the device!
+   *
+   *  For example std::vector calls this to default initialize each element. You'll segfault
+   *  if std::allocator_traits::construct tries doing that at p.
+   *
+   *  The standard is a bit confusing on this point. Implementing this is an optional requirement
+   *  of Allocator from C++11 on, its not slated to be removed.
+   *
+   *  Its deprecated for the std::allocator in c++17 and will be removed in c++20.  But we are not implementing
+   *  std::allocator.
+   *
+   *  STL containers only use Allocators through allocator_traits and std::allocator_traits handles the case
+   *  where no construct method is present in the Allocator.
+   *  But std::allocator_traits will call the Allocators construct method if present.
+   */
+  template<class U, class... Args>
+  static void construct(U* p, Args&&... args)
+  {}
+
+  /** Give std::allocator_traits something to call.
+   *  The default if this isn't present is to call p->~T() which
+   *  we can't do on device memory.
+   */
+  template<class U>
+  static void destroy(U* p)
+  {}
+
+  void copyToDevice(T* device_ptr, T* host_ptr, size_t n)
+  {
+    if(omp_target_memcpy(device_ptr, host_ptr, n, 0, 0, omp_get_default_device(), omp_get_initial_device()))
```

Comment:
Why omp_get_initial_device here? It is default device everywhere else. Please add a comment

Reply:
Initial device basically means host

Reply:
documentation of that here seems appropriate.

Reply:
omp_target_memcpy documentation has it https://www.openmp.org/spec-html/5.0/openmpsu166.html

Reply:
It is a standard use of omp_get_initial_device rather a QMCPACK unique hack. I prefer not to document in our source.

Reply:
I don't think we should be requiring a deep reading of the omp spec.  this function name is needlessly obtuse. If you don't want a comment just pull out the function calls
```
auto& device = omp_get_default_device();
auto& host = omp_get_init_device();
if(omp_target_memcpy(device_ptr, host_ptr, n, 0, 0, device, host))
```
I doubt the generated optimized code will be different.

Reply:
I like your suggestion better than adding comments.